Fix android builds requiring external (cgo) linking, but cgo is not enabled

BagToad opened the pull request:
Fixes #50

This PR proposes: 

- Extension authors will opt-in to releasing to Android / releasing to Android will be optional in future `gh-extension-precompile` versions.
- Support for building and releasing to Android targets will be best effort but cannot block the release to platforms officially supported by `gh` core. 
- Extensions using Go < 1.22 and wish to release to Android can continue to leverage older versions of `gh-extension-precompile` to avoid this new logic.
- New action inputs:
    - `release_android`
    - `android_sdk_version`
    - `android_ndk_home` 
- These changes will prompt a version bump to `gh-extension-precompile`. 

## Building on Android and change details

We will set `CGO_ENABLED=1` and `CC=${ANDROID_NDK_HOME}/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android${ANDROID_SDK_VERSION}-clang` for Android architectures, if both `release_android` and `android_sdk_version` action inputs are provided. If `release_android` is enabled but `android_sdk_version` is not set, the action will fail.

The `android_ndk_home` action input is used if the  `ANDROID_NDK_HOME` environment variable is not set. [The `$ANDROID_NDK_HOME` environment variable will be set automatically on GitHub hosted runners](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2004-Readme.md#environment-variables-2). 

When `release_android` is not set, the action will not attempt to build for Android targets, avoiding the Android specific requirements completely - this is the proposed default behavior. 

## I just want my extension to continue building for Android - what should I do?

Along with bumping the action version, modify your workflow that calls `gh-extension-precompile` to include at least `release_android` and `android_sdk_version`. 

**Example:**

```yaml
- uses: cli/gh-extension-precompile@v2
  with:
    release_android: true
    android_sdk_version: 30
    generate_attestations: true
    go_version_file: go.mod
```

If you are running the workflow on a self-hosted runner, you should confirm that either the `ANDROID_NDK_HOME` environment variable is set or that you provide the `android_ndk_home` action input.

## Testing

### Build and release no android Go@1.22
- [workflow run](https://github.com/bagtoad-enterprise-cloud-testing/gh-bagtoad-test-extension/actions/runs/10690332439/job/29634451791)
- [release & artifacts](https://github.com/bagtoad-enterprise-cloud-testing/gh-bagtoad-test-extension/releases/tag/v0.0.2)

### Build and release with android Go@1.22

- [workflow run](https://github.com/bagtoad-enterprise-cloud-testing/gh-bagtoad-test-extension/actions/runs/10690392114/job/29634654904)
- [release & artifacts](https://github.com/bagtoad-enterprise-cloud-testing/gh-bagtoad-test-extension/releases/tag/v0.0.3)

BagToad committed:
Fix android/amd64 requires external (cgo) linking, but cgo is not enabled

BagToad committed:
Change android build targets to opt-in

Extension authors will now need to opt-in to building for android targets. When opted-in to building on android, extension authors will also be required to define at least the Android SDK build-tools version.

If opted-in to building for android targets, we will attempt to build assuming Go 1.22 or newer. Previous action versions will be required to build for android using Go 1.22 and older.

BagToad committed:
Exit if release android is enabled, but android SDK version is not set

CleanCut reviewed (commented):
👍🏻 This would help my use. I have no need of android builds in the first place.

In fact, I'd love it if you could opt-in or opt-out of _all_ the platforms. I only need one.

andyfeller commented on action.yml:
```diff
@@ -20,6 +20,13 @@ inputs:
   generate_attestations:
     description: "Whether to generate artifact attestations for release binaries to establish build provenance, defaults to `false` if unspecified"
     default: false
+  release_android:
+    description: "Whether to release android binaries, defaults to `false` if unspecified"
+    default: false
+  android_ndk_home:
+    description: "Path to the Android NDK for android-amd64 and android-arm64 builds, required for Go 1.22 or newer, defaults to `ANDROID_NDK_HOME` environment variable if unspecified"
+  android_sdk_version:
+    description: "Android SDK build-tools version"
```
I think these changes deserve an update to `README.md` explaining why we are going to be cutting a `v2` update, how extension authors choosing to support Android / Termux community need to understand which Android SDK version they are targeting.

andyfeller replied:
Also, I wonder if we should update README that if you need to customize the process, then you might be on your own to copying and modifying the `build_and_release.sh` file for your needs.

BagToad replied:
I added some additional details to the readme for your review 🙇

andyfeller commented on action.yml:
```diff
@@ -20,6 +20,13 @@ inputs:
+    description: "Whether to release android binaries, defaults to `false` if unspecified"
+    default: false
+  android_ndk_home:
+    description: "Path to the Android NDK for android-amd64 and android-arm64 builds, required for Go 1.22 or newer, defaults to `ANDROID_NDK_HOME` environment variable if unspecified"
```
This is kinda long for an input description 🤔

How much of this info should be here versus on the readme?  Also, should we mention that GitHub-managed runners have this installed by default?

If we do include info on the readme, perhaps we should link to the doc sections of https://github.com/actions/runner-images for users to look up what versions come installed.

BagToad replied:
Agreed. I trimmed it down and opted to have more of this information in the readme. I added the link you suggested as well.

andyfeller commented on build_and_release.sh:
```diff
@@ -18,6 +16,17 @@ platforms=(
+  platforms+=("android-arm64")
+fi
+
+# We must know the android sdk version to build for android.
+if [[ "$RELEASE_ANDROID" == "true" && -z "$ANDROID_SDK_VERSION" ]]; then
+  echo "error: Cannot build for android without android_sdk_version." >&2
+  exit 1
+fi
```
I like we're trying to verify the state of things, should we also test that the `ANDROID_NDK_HOME` is also a directory that exists?  Or that the `cc` is a valid directory?

BagToad replied:
I think checking that `ANDROID_NDK_HOME` is a directory when `RELEASE_ANDROID` is `true` is reasonable 👍  

For `cc`, we arbitrarily clear it on L57 for every build target, so if we were to do validation on it, I think we'd need to do it based on the build target after `cc` is set in the loop. Do we still want to do that?

BagToad replied:
> I think checking that ANDROID_NDK_HOME is a directory when RELEASE_ANDROID is true is reasonable 👍

I've pushed a change to check for this 🙇

andyfeller reviewed (commented):
These changes look good, however I think there's more we need to do here before merging it especially since this will be a breaking change _(`@v2`)_

BagToad committed:
Add new details for building Go extensions on Android

BagToad committed:
Bump gh-extension-precompile to v2 in examples

BagToad committed:
doc: custom build process for Go extensions

Add heading to readme for customizing the build process for Go extensions that refers users to "extensions written in other compiled languages" section. This removes duplicating information that is the same for either case.

BagToad committed:
Slim action input description

Slim down the action input description. Lean on the readme to document this in more depth instead.

BagToad committed:
include go version file in android build examples

BagToad committed:
Link to building for android heading

BagToad committed:
Use go_version_file instead of gh_version

BagToad committed:
Update actions/checkout to v4 in readme examples

BagToad committed:
Remove extra newlines

BagToad committed:
Check that ANDROID_NDK_HOME is set

BagToad commented on README.md:
```diff
@@ -24,10 +24,10 @@ jobs:
-      - uses: actions/checkout@v3
-      - uses: cli/gh-extension-precompile@v1
+      - uses: actions/checkout@v4
+      - uses: cli/gh-extension-precompile@v2
         with:
-          go_version: "1.16"
+          go_version_file: go.mod
```
I decided to change all examples to use `go_version_file` instead of `go_version`, including this main usage example. Let me know if you disagree 😁

andyfeller commented on README.md:
```diff
@@ -36,20 +36,49 @@ When the `release` workflow finishes running, compiled binaries will be uploaded
 You can safely test out release automation by creating tags that have a `-` in them; for example: `v2.0.0-rc.1`. Such Releases will be published as _prereleases_ and will not count as a stable release of your extension.
 
-To maximize portability of built products, this action builds Go binaries with [cgo](https://pkg.go.dev/cmd/cgo) disabled. To override that, set the `CGO_ENABLED` environment variable:
+To maximize portability of built products, this action builds Go binaries with [cgo](https://pkg.go.dev/cmd/cgo) disabled unless you [enable building for Android targets](#building-for-android). To override cgo for all build targets, set the `CGO_ENABLED` environment variable:
```
Minor suggestion to avoid confusion that we enable CGO for all build targets because Android releases are enabled when CGO is only enabled for the Android build targets.

```suggestion
To maximize portability of built products, this action builds Go binaries with [cgo](https://pkg.go.dev/cmd/cgo) disabled with the exception of [Android build targets](#building-for-android). To override cgo for all build targets, set the `CGO_ENABLED` environment variable:
```

andyfeller commented on README.md:
````diff
@@ -36,20 +36,49 @@ When the `release` workflow finishes running, compiled binaries will be uploaded
 
+### Building for Android
+
+As of `gh-extension-precompile@v2`, building for Android targets like `android-arm64` and `android-amd64` is disabled by default. To enable building for Android targets, set at least the `release_android` and `android_sdk_version` action inputs:
+
+```yaml
+- uses: cli/gh-extension-precompile@v2
+  with:
+    go_version_file: go.mod
+    release_android: true
+    android_sdk_version: 34
+```
+
+If you are running the workflow on a GitHub hosted runner, you do not need to set the `android_ndk_home` input. The Android SDK Build-tools and environment variables required to build for Android targets are [pre-installed and configured on GitHub hosted runners](https://github.com/actions/runner-images/blob/8cdc506384655ceaaa62d3f800e15b844e06bea4/images/ubuntu/Ubuntu2404-Readme.md?plain=1#L214- ...
+
+However, if you are running the workflow on a self-hosted runner, you need to also configure the `android_ndk_home` action input to the installation path of the Android NDK on the runner:
+
+```yaml
+- uses: cli/gh-extension-precompile@v2
+  with:
+    go_version_file: go.mod
+    release_android: true
+    android_sdk_version: 34
+    android_ndk_home: /path/to/android-ndk
+```
````
I'm wondering how this might be simplified 🤔 

```suggestion
`gh-extension-precompile@v2` introduces a breaking change by disabling `android-arm64` and `android-amd64` build targets by default due to [Go external linking requirements](https://github.com/cli/gh-extension-precompile/issues/50#issuecomment-2078086299).

To enable Android build targets:

1. `release_android` must be set to `true`
2. `android_sdk_version` must be set to a targeted [Android API level](https://developer.android.com/tools/releases/platforms)
3. `android_ndk_home` must be set to the path to Android NDK installed on Actions runner

   `cli/gh-extension-precompile` will use pre-installed Android tools on GitHub-managed runners by default; self-hosted runners will need to install and configure this input.

   _For more information on Android NDK installed on GitHub-managed runners, see [`actions/runner-images`](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#android)
```

BagToad replied:
Love it! Thank you for seeing that opportunity to simplify what I was writing ❤️

andyfeller reviewed (commented):
99% there!  everything is looking good with minor suggestions on docs; let me know if you feels strongly and let's knock this out today

BagToad committed:
Apply suggestions from code review

Co-authored-by: Andy Feller <andyfeller@github.com>

BagToad commented:
Thank you @andyfeller 🙏 I have committed your suggestions and it's ready for another review pass ✨

BagToad commented on README.md:
```diff
@@ -36,20 +36,38 @@ When the `release` workflow finishes running, compiled binaries will be uploaded
+
+   `cli/gh-extension-precompile` will use pre-installed Android tools on GitHub-managed runners by default; self-hosted runners will need to install and configure this input.
+
+   _For more information on Android NDK installed on GitHub-managed runners, see [`actions/runner-images`](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#android)
```
Fix missing `_`

```suggestion
   _For more information on Android NDK installed on GitHub-managed runners, see [`actions/runner-images`](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#android)_
```

BagToad committed:
fix missing underscore in readme

andyfeller reviewed (approved):
![shipit-squirrel-fireworks](https://github.com/user-attachments/assets/2320321d-b1a5-47cd-8893-0d1625e27323)

Sincere thanks for unblocking our Go extension developer community while putting significant efforts to verifying these changes within Android simulator ✨ 

After merging this PR, I believe the only things that remain are:

1. Tagging the repo with `v2.0.0` tag + release, which will cause our GitHub Actions to create the `v2` branch; for more info, see below

2. Updating the release notes within the `v2` release to explain the breaking change and how users will get more information to adopt the latest changes

https://github.com/cli/gh-extension-precompile/blob/ee3e4e5b371bbd6dc7f816dc0742e86de6c64085/.github/workflows/release.yml#L11-L21

Cinderella252 reviewed (commented):
Seems to work
